purchase_cancel_qty/models/purchase_order_line.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, _
+from odoo.addons import decimal_precision as dp
+from odoo.exceptions import ValidationError
+
+PRODUCT_QTY_READONLY_STATES = {
+    'purchase': [('readonly', True)],
+    'done': [('readonly', True)],
+    'cancel': [('readonly', True)],
+}
+
+CANCELLED_QTY_EDITABLE_STATES = {
+    'purchase': [('readonly', False)],
+}
+
+
+class PurchaseOrderLine(models.Model):
+
+    _inherit = 'purchase.order.line'
+
+    product_qty = fields.Float(states=PRODUCT_QTY_READONLY_STATES)
+    ordered_qty = fields.Float(
+        string="Ordered Qty", readonly=True, copy=False,
+        digits=dp.get_precision('Product Unit of Measure'))
+    cancelled_qty = fields.Float(
+        string="Cancelled Qty", readonly=True, copy=False,
+        inverse='_inverse_cancelled_qty',
+        track_visibility='onchange',
+        digits=dp.get_precision('Product Unit of Measure'),
+        states=CANCELLED_QTY_EDITABLE_STATES)
+
+    _sql_constraints = [
+        (
+            'cancelled_qty_gt_zero',
+            'check(COALESCE(cancelled_qty, 0) >= 0)',
+            _("Cancelled quantity must be greater or equal to 0.")
```

Comment:
Why not do this in `_check_cancelled_qty`?

Reply:
To prevent users to bypass the ORM, but in this case, I should have done the other check (cancelled_qty <= ordered_qty) as a sql constraint too.

Reply:
I tend to prefer python constraints because they allow for more explicit error messages.